### PR TITLE
Improve serialization to fix #4877

### LINKED ISSUE #4877: When saving: The libary has been modified by another program
Edit:
JabRef 5.0, 5.1, 5.2, 5.4 till current version (Nov2021) is affected.

If you add a comment please add
- your operating system
- encoding of the document (e.g. ASCII, ANSI, ISO-8859, UTF-16 , ...)
- the line-seperators used in the file (lineforward or cariage-return or both or mixed)
- the settings of `Options` > `Preferences` > `File`
-- (a) Newline seperator (try LineForward)
-- (b) Always reformat BIB file on save and export (try disabling)
-- (c) Autosave local libraries (try disabling)

----
Original post

    JabRef 5.0-dev--snapshot--2019-04-09--master--0dd091f31
    Linux 4.15.0-47-generic amd64 
    Java 1.8.0_201

Ubuntu 18.04

- [x] I have tested the latest development version from http://builds.jabref.org/master/ and the problem persists

Steps to reproduce the behavior: (same as #4810 )
1. open .bib-file
2. press "ctrl" & "s"
3. Repeat step 2 till Bug appears.
4. The message "The library has been modified by another program" appears
5. DO NOT accept the changes, otherwise all entries will be doubled (click dismiss)
Screenshot from 2019-04-09 16-57-25
Screenshot from 2019-04-09 16-57-34

<details>
  <summary>Log File</summary>
  
  ```
Opening: /home/jkalliau/Desktop/Jabrefissue/LargeFile.bib
Can only load style files: Preview
Vorschaustil geändert zu: Vorschau
Speichere Bibliothek...
Opening: /tmp/jabref2591679195168273738.bib
Opening: /home/jkalliau/Desktop/Jabrefissue/LargeFile.bib
Bibliothek gespeichert '/home/jkalliau/Desktop/Jabrefissue/LargeFile.bib'.
  ```
</details>

<details>
  <summary>Terminal</summary>
  
  ```
16:45:54.405 [JavaFX Application Thread] INFO  org.jabref.logic.importer.OpenDatabase - Opening: /home/jkalliau/Desktop/Jabrefissue/LargeFile.bib
16:45:54.963 [JavaFX Application Thread] ERROR org.jabref.logic.citationstyle.CitationStyle - Can only load style files: Preview
16:45:55.220 [JavaFX Application Thread] INFO  org.jabref.gui.FXDialogService - Vorschaustil geändert zu: Vorschau
16:46:06.495 [JavaFX Application Thread] INFO  org.jabref.gui.FXDialogService - Speichere Bibliothek...
16:46:06.524 [pool-3-thread-2] INFO  org.jabref.logic.importer.OpenDatabase - Opening: /tmp/jabref2591679195168273738.bib
16:46:06.526 [pool-3-thread-2] INFO  org.jabref.logic.importer.OpenDatabase - Opening: /home/jkalliau/Desktop/Jabrefissue/LargeFile.bib
16:46:06.530 [JavaFX Application Thread] INFO  org.jabref.gui.FXDialogService - Bibliothek gespeichert '/home/jkalliau/Desktop/Jabrefissue/LargeFile.bib'.
16:53:39.039 [JavaFX Application Thread] INFO  org.jabref.gui.FXDialogService - In die Zwischenablage kopiert
  ```
</details>

### PR DESCRIPTION
**Pull request by tobiasdiez**
Hopefully, fixes #4877 for good. Reason for the remaining issue is described in https://github.com/JabRef/jabref/issues/4877#issuecomment-573701393.
This is fixed in this PR by:
- Consistently remember in the metadata which encoding was used when saving
- The library mode is already determined (if not set explicitly by the user - which should be the normal case) upon import/loading instead of at an arbitrary point when the mode is first used.
- No longer format field values upon writing, instead the written field values coincide exactly with the field values contained in the entry. The little pre-save processing (newlines, trim values) is now implemented as a mandatory save action.

Along the way, I refactored a bit of code (biggest change: remove Defaults class) and fixed very small code issues.

- [ ] Change in CHANGELOG.md described (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for bigger UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not: Issue created at <https://github.com/JabRef/user-documentation/issues>.

### CODE REVIEW

**Comment by tobiasdiez** [comment]
The `LatexFieldFormatter` was actually only used for writing normal fields. I also thought that it pre-formats the field value, but it turns out that it actually contained all the logic surrounding bibtex-strings (i.e. brackets needed or not) and a bit of pre-processing. This is why I renamed it now to `FieldWriter` which should better reflect what this class does.

File: src/test/java/org/jabref/logic/bibtex/BibEntryWriterTest.java [hunk 1]
```diff
@@ -515,24 +515,6 @@ void doNotWriteEmptyFields() throws IOException {
         assertEquals(expected, actual);
     }
 
-    @Test
-    void trimFieldContents() throws IOException {
```
Comment (koppor) [inline src/test/java/org/jabref/logic/bibtex/BibEntryWriterTest.java]: Can we keep this test --> maybe for the whitespace removal thing?

**Comment by koppor** [comment]
I think, if the .bib file has `LF` endings, multine fields will be written using `CRLF` on Windows. - Should we add the line feed property to the meta data? Or just convert the bib file completely to LF upon write (not recommended)

File: src/test/java/org/jabref/logic/bibtex/BibEntryWriterTest.java (hunk 1 above)
Comment (tobiasdiez) [reply src/test/java/org/jabref/logic/bibtex/BibEntryWriterTest.java]: I moved the test to the database writer as this is the new place where the field content is trimmed.
https://github.com/JabRef/jabref/pull/5838/files#diff-128297a947efc0b9e025fd5541c72376R650

**Comment by tobiasdiez** [comment]
Thanks for the quick (and positive) reviews. I'll merge now.

I didn't change the logic of the newline normalization (just the place where it occurs). I guess it might be indeed a good idea to make the line ending configurable per database - on the other hand, no user complained so far.

**Comment by koppor** [comment]
Because the blame git for issues ^^. Saw that often. I also wondered when opening some bib files in Notepad++, why we have mixed line endings. Now I know it. Hope, someone will work on the fix. It is so annoying if git complains about line endings and one does not know why.

**Comment by wujastyk** [comment]
Sorry, I just made [this comment](https://github.com/JabRef/jabref/issues/4877#issuecomment-583054261) under #4877, which is probably the wrong place to have made it.

**Comment by koppor** [comment]
Wrong line endings reported at https://github.com/koppor/jabref/issues/390

**Pull request merged**